python/summarize_sgfs.py

```diff
@@ -0,0 +1,295 @@
+    elif (winner == 'w'):
+      self.results[(pla_black, pla_white)].lost += 1
+    else:
+      self.results[(pla_black, pla_white)].draw += 1
+
+  def _estimate_elo(self) -> elo.EloInfo:
+    """Estimate and print elo values. This function must be called after add all the sgfs/sgf files"""
+    pla_names = set(itertools.chain(*(name_pair for name_pair in self.results.keys())))
+    data = []
+    for pla_Balck in pla_names:
+      for pla_White in pla_names:
+        if (pla_Balck == pla_White):
+          continue
+        else:
+          record = self.results[(pla_Balck, pla_White)]
+          total = record.win + record.lost + record.draw
+          win = record.win + 0.5 * record.draw
+          winrate = win / total
+          data.extend(elo.likelihood_of_games(pla_Balck, pla_White, total, winrate, False))
+    data.extend(elo.make_center_elos_prior(list(pla_names), self._elo_prior))
+    info = elo.compute_elos(data, verbose=True)
+    return info
+
+  def _print_player_info(self):
+    pla_names = set(itertools.chain(*(name_pair for name_pair in self.results.keys())))
+    max_len = len(max(pla_names, key=len))
+    title = 'Player Name'
+    title_space = max(len(title), max_len) + 1
+    print(f"{title:<{title_space}}: Player ID")
+    idx = 0
+    for pla in pla_names:
+      print(f"{pla:<{title_space}}: {idx}")
+      idx += 1
+
+  def _print_result_matrix(self):
+    results_matrix = self._build_result_matrix()
+    max_game_played_space = int(math.log10(max([max(sublist) for sublist in results_matrix]))) + 5
+    row_format = f"{{:>{max_game_played_space}}}" * (len(results_matrix) + 1)
+    print(row_format.format("", *range(len(results_matrix))))
+    for playerID, row in zip(range(len(results_matrix)), results_matrix):
+      print(row_format.format(playerID, *row))
+
+  def _build_result_matrix(self) -> list:
+    pla_names = list(set(itertools.chain(*(name_pair for name_pair in self.results.keys()))))
+    results_matrix = []
+    for pla1 in pla_names:
+      row = []
+      for pla2 in pla_names:
+        if (pla1 == pla2):
+          row.append(0)
+          continue
+        else:
+          pla1_pla2 = self.results[(pla1, pla2)]
+          pla2_pla1 = self.results[(pla2, pla1)]
+          win = pla1_pla2.win + pla2_pla1.lost + 0.5 * (pla1_pla2.draw + pla2_pla1.draw)
+          row.append(win)
+      results_matrix.append(row)
+    return results_matrix
+
+
+if __name__ == "__main__":
+  description = """
+  Summarize SGF/SGFs files and estimate Bayes Elo score for each of the player.
+  """
+  parser = argparse.ArgumentParser(description=description)
+  parser.add_argument('-input-dirs', help='sgf/sgfs files input directories. If multiple directories, seperate them '
+                                          'with a space', required=True,
+                      nargs='+')
+  parser.add_argument('-search-subdir', help='If we also need to search all sub-directories', required=False,
```

Comment:
Just a question - is there a specific reason for making this an option, instead of just always searching all the subdirectories?

Reply:
In the 3rd page of the paper on MM algorithm for Bayes Elo calculation. It required that the players have to play with each other with win and lost, otherwise, the algorithm can't calculate the Elo score. Sometimes, if they keep searching all the subdirectories, it may accidentally include some games with players who didn't play with all the others, and only played with with a group, and this group is isolated. So the calculation will report errors. So we put `search-subdir` as false unless users know what they want to do.

Reply:
This sort of makes sense, but why do you have directories where including all the sgfs in the directory are ones you want to include, but somehow you have further subdirectories that include extraneous data that should not be included? And you specifically rely on the extraneous files to be in a subdirectory rather than in the top-level directory to be included? This seems like a confusing way to organize your SGF files.

Reply:
That is correct comment. When I was trying to debug my codes and test the subdirectories function, I copied some other sgfs files into a subdirectory and found this error. So we made the search-sub default as false.